### PR TITLE
RATIS-1663. Record call id for board casting a heartbeat.

### PR DESCRIPTION
**Pull request by szetszwo**
See https://issues.apache.org/jira/browse/RATIS-1663

### CODE REVIEW

File: ratis-common/src/main/java/org/apache/ratis/rpc/CallId.java [hunk 1]
```diff
@@ -27,6 +28,17 @@
 public final class CallId {
   private static final AtomicLong CALL_ID_COUNTER = new AtomicLong();
 
+  private static final Comparator<Long> COMPARATOR = (left, right) -> {
+    final long diff = left - right;
+    // check diff < Long.MAX_VALUE/2 for the possibility of numerical overflow.
+    return diff == 0? 0: diff > 0 && diff < Long.MAX_VALUE/2? 1: -1;
```
Comment (codings-dan) [inline ratis-common/src/main/java/org/apache/ratis/rpc/CallId.java:34]: Why this `COMPARATOR` is different from GrpcLogAppender#CALL_ID_COMPARATOR?

File: ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java [hunk 2]
```diff
@@ -56,9 +57,16 @@
 public class GrpcLogAppender extends LogAppenderBase {
   public static final Logger LOG = LoggerFactory.getLogger(GrpcLogAppender.class);
 
+  private static final Comparator<Long> CALL_ID_COMPARATOR = (left, right) -> {
+    // calculate diff in order to take care the possibility of numerical overflow,
```
Comment (codings-dan) [inline ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java:61]: typo , -> .

File: ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    private final LogAppender appender;
+    private final long minCallId;
+    private volatile boolean acknowledged = false;
+
+    HeartbeatAck(LogAppender appender) {
+      this.appender = appender;
+      this.minCallId = appender.getCallId();
+    }
+
+    boolean isAcknowledged() {
```
Comment (codings-dan) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java:35]: Could you add a comment for this method?

**Review by codings-dan — COMMENTED** [review]
@szetszwo Thanks for working on this, just have a question, thank you!

File: ratis-common/src/main/java/org/apache/ratis/rpc/CallId.java (hunk 1 above)
Comment (szetszwo) [reply ratis-common/src/main/java/org/apache/ratis/rpc/CallId.java:34]: It generates call IDs in a different way -- this only allows >=0 values but GrpcLogAppender allows +/- values.

File: ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java (hunk 2 above)
Comment (szetszwo) [reply ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java:61]: Let me simply remove it.  Thanks.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java (hunk 3 above)
Comment (szetszwo) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java:37]: Done.

File: ratis-common/src/main/java/org/apache/ratis/rpc/CallId.java (hunk 1 above)
Comment (codings-dan) [reply ratis-common/src/main/java/org/apache/ratis/rpc/CallId.java:34]: Got it, thanks!

**Review by codings-dan — APPROVED** [review]
The change looks good, +1

**Comment by szetszwo** [comment]
@codings-dan , thanks a lot for reviewing this!

**Pull request merged**